File: compiler/bench/anagram.fs
\ expected: 50000
\ Anagram detection benchmark - check anagram pairs, run 50000 times

create word1 32 allot
create word2 32 allot
create counts1 26 allot
create counts2 26 allot

: init-words ( -- )
  s" listen" drop word1 6 cmove
  s" silent" drop word2 6 cmove ;

: clear-counts ( -- )
  counts1 26 0 fill
  counts2 26 0 fill ;

: count-char ( c counts -- )
  swap [char] a - + dup c@ 1+ swap c! ;

: count-word ( addr len counts -- )
  -rot 0 do
    over i + c@ over count-char
  loop 2drop ;

: compare-counts ( -- flag )
  -1
  26 0 do
    counts1 i + c@ counts2 i + c@ <> if drop 0 leave then
  loop ;

: anagram? ( -- flag )
  clear-counts
  word1 6 counts1 count-word
  word2 6 counts2 count-word
  compare-counts ;

: main ( -- )
  init-words
  0
  50000 0 do
    anagram? if 1+ then
  loop
  . cr ;

File: compiler/bench/anagram.c
// expected: 50000
// Anagram detection benchmark - check anagram pairs, run 50000 times

#include <stdio.h>
#include <string.h>

char word1[32];
char word2[32];
int counts1[26];
int counts2[26];

void init_words(void) {
    strcpy(word1, "listen");
    strcpy(word2, "silent");
}

void clear_counts(void) {
    memset(counts1, 0, sizeof(counts1));
    memset(counts2, 0, sizeof(counts2));
}

void count_word(const char *s, int len, int *counts) {
    for (int i = 0; i < len; i++) {
        counts[s[i] - 'a']++;
    }
}

int compare_counts(void) {
    for (int i = 0; i < 26; i++) {
        if (counts1[i] != counts2[i]) return 0;
    }
    return 1;
}

int anagram(void) {
    clear_counts();
    count_word(word1, 6, counts1);
    count_word(word2, 6, counts2);
    return compare_counts();
}

int main(void) {
    init_words();
    int result = 0;
    for (int i = 0; i < 50000; i++) {
        if (anagram()) result++;
    }
    printf("%d\n", result);
    return 0;
}
